domain/application/resource/resourcestore.go

```diff
@@ -9,25 +9,35 @@ import (
 	"sync"
 
 	"github.com/juju/juju/core/objectstore"
+	coreresource "github.com/juju/juju/core/resources"
 	"github.com/juju/juju/domain/application/errors"
 	"github.com/juju/juju/internal/charm/resource"
 )
 
 // ResourceStore provides a list of methods necessary for interacting with
 // a store for the resource.
 type ResourceStore interface {
-	// Get returns an io.ReadCloser for data at path, namespaced to the
-	// model.
-	Get(context.Context, string) (io.ReadCloser, int64, error)
-}
+	// Get returns an io.ReadCloser for a resource in the resource store.
+	Get(
+		ctx context.Context,
+		resourceUUID coreresource.ID,
```

Comment:
I believe conflating the UUID from the resource table with the storage path of said resource is not a good idea. Especially as it is not well documented in the code nor the PR description.

Let's continue to use `application-<app-name>/resources/<resource-name>` as the storage-path until it becomes apparent that's a bad idea.

Reply:
I just did some validation and this is actually currently broken in 4.0 with the object store vs. 3.6. When you upload a file with the same name twice in 4.0 you get an error back from the object store because the path already exists. In 3.6 each file is assigned a unique id by the blob store and the `storedResources` table keeps track of which is which. So you can attach multiple `foo.txt` the resource is updated with the latest one. 

If my understanding of the current/proposed DDL is correct, we will have a unique UUID/entry in the `resource` table for every resource uploaded/revision added (let me know if this is an incorrect assumption!). When adding a resource, we can generate the UUID, use it to put the resource in the object store/oci image metadata store, then add the resource info with that UUID to the `resource` table in state.

Reply:
It's not clear that every uploaded resource will get a new resource UUID right now. 

Choosing to use the resource uuid as the storage path is an implementation detail. This does not equate to codifing that detail. 

The changes in #18388 indicate using the resource uuid to get the object store metadata uuid from the resource_file_store table, which can be used to get the storage path from the object_store_metadata_path table.

It would be easier if the object store Get method used an object store UUID for sure, but that won't change for a while.

Reply:
Discussed this with @manadart. The UUID can be used as the storage path because we can generate the UUID when a resource is uploaded, put it in storage, then create the row in `resource`.
Because the same version of the resource is used regardless of the charm version (i.e. when a charm is updated the charm_resource will also be updated), we will only be changing the resource the application uses in `application_resource` when the resource in the store also needs changing. The exception to this would be a user uploaded resource, but that can be handled as part of charm upgrades.

Reply:
> The exception to this would be a user uploaded resource, but that can be handled as part of charm upgrades.

@Aflynn50 this statement is only partially true. `juju attach-resource` must also handle uploaded resources. That functionality is a requirement of the resource epic. The code paths are the same for `attach-resource`, `deploy`, and `refresh` for uploading local resources. We need to handle that now.

Reply:
For attach-resource this isn't a problem, if we have a new blob to store we will get a new resource UUID can use that for storing the blob.

The issue with a charm being upgraded is that if say the `path` field in the `charm_resource` table changes and you have a user uploaded resource, you would want to keep the user uploaded resource, but add a new row in the resource table with the new charm_uuid and remove the old one.

I realise this wasn't very clear in the message above!